constantine/math/arithmetic/assembly/limbs_asm_mul_mont_arm64.nim

```diff
@@ -0,0 +1,444 @@
+    ctx.umulh u, lhs, rhs
+    ctx.adds dst, addend, u
+    swap(u, v)
+  template mulhiadd_cio(ctx, dst, lhs, rhs, addend) {.dirty.} =
+    ctx.umulh u, lhs, rhs
+    ctx.adcs dst, addend, u
+    swap(u, v)
+  template mulhiadd_ci(ctx, dst, lhs, rhs, addend) {.dirty.} =
+    ctx.umulh u, lhs, rhs
+    ctx.adc dst, addend, u
+    swap(u, v)
+
+  result.add quote do:
+    static: doAssert: sizeof(SecretWord) == sizeof(ByteAddress)
+
+    var `tsym`{.noInit, used.}: typeof(`r_PIR`)
+    # Assumes 64-bit limbs on 64-bit arch (or you can't store an address)
+    var `scratchSym` {.noInit.}: Limbs[`scratchSlots`]
+    `scratchSym`[0] = cast[SecretWord](`a_PIR`[0][0].unsafeAddr)
+    `scratchSym`[1] = cast[SecretWord](`b_PIR`[0][0].unsafeAddr)
+
+    var `akSym` {.noInit.}: typeof(`a_PIR`[0])
+
+  # Algorithm
+  # -----------------------------------------
+  # for i=0 to N-1
+  #   tN := 0
+  #   for k=0 to K-1
+  #     A := 0
+  #     for j=0 to N-1
+  # 		  (A,t[j])  := t[j] + a[k][j]*b[k][i] + A
+  #     tN += A
+  #   m := t[0]*m0ninv mod W
+  # 	C,_ := t[0] + m*M[0]
+  # 	for j=1 to N-1
+  # 		(C,t[j-1]) := t[j] + m*M[j] + C
+  #   t[N-1] = tN + C
+
+  for i in 0 ..< N:
+    # Multiplication step
+    # -------------------------------
+    ctx.comment "  Multiplication step"
+    ctx.comment "  tN = 0"
+    ctx.mov tN, xzr
+    for k in 0 ..< K:
+      ctx.comment "    A = 0"
+      ctx.mov A, xzr
+
+      ctx.comment "    bi <- b[k, i]"
+      ctx.ldr bi, b[k, i]
+
+      ctx.comment "    load a[k] in registers"
+      let lastEven = N.round_step_down(2)
+      for i in countup(0, lastEven-1, 2):
+        ctx.ldp ak[i], ak[i+1], a[k, i]
+      if lastEven != N:
+        ctx.ldr ak[N-1], a[k, N-1]
+
+      ctx.comment "    (A,t[0])  := t[0] + a[k][0]*b[k][i] + A"
+      if k == 0 and i == 0: # First accumulation, overwrite t[0]
+        for j in 0 ..< N:
+          ctx.mul t[j], ak[j], bi
+      else:
+        ctx.mulloadd_co(t[0], ak[0], bi, t[0])
+        for j in 1 ..< N:
+          ctx.mulloadd_cio(t[j], ak[j], bi, t[j])
+        ctx.adc A, xzr, xzr                        # assumes N > 1
+
+      ctx.mulhiadd_co(t[1], ak[0], bi, t[1])       # assumes N > 1
+      for j in 2 ..< N:
+        ctx.mulhiadd_cio(t[j], ak[j-1], bi, t[j])
+      ctx.mulhiadd_ci(A, ak[N-1], bi, A)
+
+      ctx.add tN, tN, A
+
+    # Reduction step
+    # -------------------------------
+    ctx.comment "  Reduction step"
+
+    ctx.mul m, t[0], m0ninv
+    ctx.mul u, m, M[0]
+    ctx.cmn t[0], u         # TODO: bad latency chain, hopefully done parallel to prev loop
+    swap(u, v)
+
+    for j in 1 ..< N:
+      ctx.mulloadd_cio(t[j-1], m, M[j], t[j])
+    ctx.adc t[N-1], tN, xzr
+
+    # assumes N > 1
+    ctx.mulhiadd_co(t[0], m, M[0], t[0])
+    for j in 1 ..< N-1:
+      ctx.mulhiadd_cio(t[j], m, M[j], t[j])
+    ctx.mulhiadd_ci(t[N-1], m, M[N-1], t[N-1])
+
+
+  if lazyReduce:
+    for i in 0 ..< N:
+      ctx.str t[i], r[i]
+  else:
+    # Final substraction
```

Comment:
Final subtraction to refactor in dedicated proc common with field addition and Montgomery reduction